group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -5524,5 +5670,2897 @@ private static Record newGroupMetadataRecord(
+            group,
+            COMPLETING_REBALANCE,
+            Collections.singleton("follower-instance-id")
+        );
+    }
+
+    @Test
+    public void testStaticMemberRejoinAsFollowerWithUnknownMemberId() throws Exception {
+        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
+            .build();
+        GenericGroup group = context.createGenericGroup("group-id");
+
+        RebalanceResult rebalanceResult = staticMembersJoinAndRebalance(
+            context,
+            "group-id",
+            "leader-instance-id",
+            "follower-instance-id"
+        );
+
+        // A static follower rejoin with no protocol change will not trigger rebalance.
+        JoinGroupRequestData request = new JoinGroupRequestBuilder()
+            .withGroupId("group-id")
+            .withGroupInstanceId("follower-instance-id")
+            .withMemberId(UNKNOWN_MEMBER_ID)
+            .withProtocolSuperset()
+            .build();
+
+        CompletableFuture<JoinGroupResponseData> followerJoinFuture = new CompletableFuture<>();
+        CoordinatorResult<Void, Record> result = context.sendGenericGroupJoin(
+            request,
+            followerJoinFuture,
+            true,
+            true);
+
+        assertEquals(
+            Collections.singletonList(newGroupMetadataRecord(group, MetadataVersion.latest())),
+            result.records()
+        );
+        // Simulate successful write to log.
+        result.appendFuture().complete(null);
+
+        assertTrue(followerJoinFuture.isDone());
+
+        // Old leader shouldn't be timed out.
+        assertTrue(group.hasStaticMember("leader-instance-id"));
+
+        JoinGroupResponseData expectedFollowerResponse = new JoinGroupResponseData()
+            .setErrorCode(Errors.NONE.code())
+            .setGenerationId(rebalanceResult.generationId) // The group has not changed.
+            .setMemberId(followerJoinFuture.get().memberId())
+            .setLeader(rebalanceResult.leaderId)
+            .setProtocolName("range")
+            .setProtocolType("consumer")
+            .setSkipAssignment(false)
+            .setMembers(Collections.emptyList());
+
+        checkJoinGroupResponse(
+            expectedFollowerResponse,
+            followerJoinFuture.get(),
+            group,
+            STABLE,
+            Collections.emptySet()
+        );
+        assertNotEquals(rebalanceResult.followerId, followerJoinFuture.get().memberId());
+
+        CompletableFuture<SyncGroupResponseData> syncFuture = new CompletableFuture<>();
+        SyncGroupRequestData syncRequest = new SyncGroupRequestBuilder()
+            .withGroupId("group-id")
+            .withGroupInstanceId("follower-instance-id")
+            .withGenerationId(rebalanceResult.generationId)
+            .withMemberId(followerJoinFuture.get().memberId())
+            .build();
+
+        result = context.sendGenericGroupSync(syncRequest, syncFuture);
```

Comment:
This code is quite repeated as well. I wonder if we could do something like this:

```
SyncResult result = context.genericGroupSync(new SyncGroup....));
```

`SyncResult` would have two attributes: response (the future) and records.